lib/ansible/utils/collection_loader.py

```diff
@@ -0,0 +1,305 @@
+            pkg_name = p[0]
+            pkg_def = p[1]
+
+            newmod = ModuleType(pkg_name)
+            newmod.__package__ = pkg_name
+            newmod.__file__ = '<ansible_synthetic_collection_package>'
+            pkg_type = pkg_def.get('type')
+
+            # TODO: need to rethink map style so we can just delegate all the loading
+
+            if pkg_type == 'flatmap':
+                newmod.__loader__ = AnsibleFlatMapLoader(import_module(pkg_def['flatmap']))
+            newmod.__path__ = []
+
+            sys.modules[pkg_name] = newmod
+
+
+    @property
+    def _collection_paths(self):
+        return self._playbook_paths + self._configured_paths
+
+    def set_playbook_paths(self, playbook_paths):
+        if isinstance(playbook_paths, string_types):
+            playbook_paths = [playbook_paths]
+
+        added_paths = set()
+
+        # de-dupe and ensure the paths are native strings (Python seems to do this for package paths etc, so assume it's safe)
+        self._playbook_paths = [to_native(p) for p in playbook_paths if not (p in added_paths or added_paths.add(p))]
+        # FIXME: only allow setting this once, or handle any necessary cache/package path invalidations internally?
+
+    def find_module(self, fullname, path=None):
+        # this loader is only concerned with items under the Ansible Collections namespace hierarchy, ignore others
+        if fullname.startswith('ansible_collections.') or fullname == 'ansible_collections':
+            return self
+
+        return None
+
+    def load_module(self, fullname):
+        if sys.modules.get(fullname):
+            return sys.modules[fullname]
+
+        # this loader implements key functionality for Ansible collections
+        # * implicit distributed namespace packages for the root Ansible namespace (no pkgutil.extend_path hackery reqd)
+        # * implicit package support for Python 2.7 (no need for __init__.py in collections, except to use standard Py2.7 tooling)
+        # * preventing controller-side code injection during collection loading
+        # * (default loader would execute arbitrary package code from all __init__.py's)
+
+        parent_pkg_name = '.'.join(fullname.split('.')[:-1])
+
+        parent_pkg = sys.modules.get(parent_pkg_name)
+
+        if parent_pkg_name and not parent_pkg:
+            raise ImportError('parent package {0} not found'.format(parent_pkg_name))
+
+        # are we at or below the collection level? eg a.mynamespace.mycollection.something.else
+        # if so, we don't want distributed namespace behavior; first mynamespace.mycollection on the path is where
+        # we'll load everything from (ie, don't fall back to another mynamespace.mycollection lower on the path)
+        sub_collection = fullname.count('.') > 1
+
+        synpkg_def = _SYNTHETIC_PACKAGES.get(fullname)
+
+        # FIXME: collapse as much of this back to on-demand as possible (maybe stub packages that get replaced when actually loaded?)
```

Comment:
Might be useful to extract each of the module constructing clauses below into their own methods. 
ie, one for 'map', 'flatmap', and 'pkg_only'. 

Among other things, that would make it easier to unit test each of those.